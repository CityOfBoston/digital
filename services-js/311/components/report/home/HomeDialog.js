// @flow

import React from 'react';
import { css } from 'glamor';
import { action } from 'mobx';
import { observer } from 'mobx-react';
import Head from 'next/head';

import type { AppStore } from '../../../data/store';

import { MEDIA_LARGE } from '../../style-constants';
import FormDialog from '../../common/FormDialog';
import SectionHeader from '../../common/SectionHeader';
import DescriptionBox from '../../common/DescriptionBox';
import ServiceList from './ServiceList';

export type Props = {
  store: AppStore,
  routeToServiceForm: (code: ?string) => void,
  // eslint-disable-next-line react/no-unused-prop-types
  stage: 'home' | 'service',
};

const DESCRIPTION_HEADER_STYLE = css({
  display: 'none',
  [MEDIA_LARGE]: {
    display: 'block',
  },
});

const NEXT_BUTTON_STYLE = css({
  width: '100%',
  [MEDIA_LARGE]: {
    display: 'none',
  },
});

const SERVICE_PICKER_STYLE = css({
  display: 'none',
  [MEDIA_LARGE]: {
    display: 'block',
  },
});

function renderHome({ store, routeToServiceForm }: Props) {
  return (
    <FormDialog>
      <Head>
        <title>BOS:311 — Report a Problem</title>
      </Head>

      <SectionHeader>311: Boston City Services</SectionHeader>

      <div className="m-v500">
        <div className="g">
          <div className="g--8">
            <h3 className={`stp m-v300 ${DESCRIPTION_HEADER_STYLE.toString()}`}>
              <span className="stp-number">1</span>
              What can we do for you?
            </h3>

            <DescriptionBox
              minHeight={222}
              maxHeight={222}
              text={store.requestForm.description}
              placeholder="How can we help?"
              onInput={action((ev) => { store.requestForm.description = ev.target.value; })}
            />
          </div>

          <div className={`g--44 ${SERVICE_PICKER_STYLE.toString()}`}>
            <h3 className="stp m-v300">Top Service Requests</h3>
            <div style={{ height: 222, overflowY: 'scroll' }}>
              <ServiceList serviceSummaries={store.serviceSummaries} onServiceChosen={routeToServiceForm} />
            </div>
          </div>
        </div>

      </div>

      <button className={`btn ${NEXT_BUTTON_STYLE.toString()}`} onClick={() => { routeToServiceForm(); }}>Next</button>
    </FormDialog>
  );
}

function renderServicePicker({ store, routeToServiceForm }: Props) {
  return (
    <FormDialog>
      <Head>
        <title>BOS:311 — Report a Problem</title>
      </Head>

      <SectionHeader>311: Boston City Services</SectionHeader>

      <div className="m-v500">
        <div className="t--intro">Pick the most related:</div>
        <div style={{ height: 222, overflowY: 'scroll' }}>
          <ServiceList serviceSummaries={store.serviceSummaries} onServiceChosen={routeToServiceForm} />
        </div>
      </div>
    </FormDialog>
  );
}

export default observer(function HomeDialog(props: Props) {
  const { stage } = props;
  switch (stage) {
    case 'home': return renderHome(props);
    case 'service': return renderServicePicker(props);
    default: return null;
  }
});
