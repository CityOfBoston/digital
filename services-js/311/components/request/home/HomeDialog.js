// @flow

import * as React from 'react';
import { action, observable, reaction, runInAction } from 'mobx';
import { observer } from 'mobx-react';
import debounce from 'lodash/debounce';
import type { Context } from 'next';
import Router from 'next/router';

import type { RequestAdditions } from '../../../server/next-handlers';

import type { ServiceSummary } from '../../../data/types';
import type { AppStore, LanguagePreference } from '../../../data/store';
import makeLoopbackGraphql from '../../../data/dao/loopback-graphql';
import type { LoopbackGraphql } from '../../../data/dao/loopback-graphql';
import loadServiceSuggestions from '../../../data/dao/load-service-suggestions';
import loadTopServiceSummaries from '../../../data/dao/load-top-service-summaries';

import FormDialog from '../../common/FormDialog';
import LoadingIcons from '../../common/LoadingIcons';
import { CENTERED_DIALOG_STYLE } from '../../style-constants';
import TranslateDialog from '../translate/TranslateDialog';

import HomePane from './HomePane';
import ChooseServicePane from './ChooseServicePane';

type Stage = 'home' | 'choose';

export type InitialProps = {|
  topServiceSummaries: ServiceSummary[],
  description: string,
  stage: Stage,
  bypassTranslateDialog: boolean,
|};

export type Props = {|
  store: AppStore,
  loopbackGraphql: LoopbackGraphql,
  ...InitialProps,
|};

@observer
export default class HomeDialog extends React.Component<Props> {
  @observable description: string;

  @observable.shallow suggestedServiceSummaries: ?(ServiceSummary[]) = null;
  serviceSuggestionsDisposer: ?Function;

  // Called by RequestLayout
  static async getInitialProps({
    query,
    req,
  }: Context<RequestAdditions>): Promise<InitialProps> {
    const { stage, description, translate } = query;
    const loopbackGraphql = makeLoopbackGraphql(req);

    return {
      topServiceSummaries: await loadTopServiceSummaries(loopbackGraphql, 5),
      stage: stage === 'choose' ? stage : 'home',
      description: description || '',
      bypassTranslateDialog: translate === '0',
    };
  }

  @action
  componentWillMount() {
    this.description = this.props.description;
  }

  componentDidMount() {
    this.serviceSuggestionsDisposer = reaction(
      () => this.description,
      (description: string) => {
        if (description === '') {
          this.suggestedServiceSummaries = [];
        } else {
          this.suggestedServiceSummaries = null;
          this.suggestServices(description);
        }
      },
      {
        fireImmediately: true,
        name: 'update service suggestions from description',
      }
    );

    LoadingIcons.preload();
  }

  componentWillUnmount() {
    if (this.serviceSuggestionsDisposer) {
      this.serviceSuggestionsDisposer();
    }
  }

  suggestServices = debounce(async (description: string) => {
    try {
      const suggestedServiceSummaries = await loadServiceSuggestions(
        this.props.loopbackGraphql,
        description
      );

      runInAction('suggestServices result', () => {
        this.suggestedServiceSummaries = suggestedServiceSummaries;
      });
    } catch (e) {
      runInAction('suggestServices error', () => {
        this.suggestedServiceSummaries = [];
      });
      throw e;
    }
  }, 500);

  @action.bound
  handleDescriptionChanged(ev: SyntheticInputEvent<>) {
    this.description = ev.target.value;
  }

  @action.bound
  routeToChoose() {
    Router.push(
      `/request?stage=choose&description=${encodeURIComponent(
        this.description
      )}`,
      '/request'
    );
  }

  render() {
    const { stage, store, bypassTranslateDialog } = this.props;
    const languages: LanguagePreference[] = store.languages;

    const translateLanguage = TranslateDialog.findLanguage(languages);
    const showTranslate =
      translateLanguage &&
      translateLanguage !== 'en' &&
      !bypassTranslateDialog &&
      stage === 'home';

    const narrow = stage === 'choose';

    if (showTranslate) {
      return <TranslateDialog languages={languages} showContinueInEnglish />;
    }

    return (
      <div className={CENTERED_DIALOG_STYLE}>
        <FormDialog narrow={narrow} noPadding>
          {stage === 'home' && this.renderHome()}
          {stage === 'choose' && this.renderServicePicker()}
        </FormDialog>
      </div>
    );
  }

  renderHome() {
    const { topServiceSummaries, store } = this.props;
    return (
      <HomePane
        store={store}
        description={this.description}
        handleDescriptionChanged={this.handleDescriptionChanged}
        nextFn={this.routeToChoose}
        topServiceSummaries={topServiceSummaries}
      />
    );
  }

  renderServicePicker() {
    const { store: { ui } } = this.props;
    return (
      <ChooseServicePane
        description={this.description}
        suggestedServiceSummaries={this.suggestedServiceSummaries}
        ui={ui}
      />
    );
  }
}
