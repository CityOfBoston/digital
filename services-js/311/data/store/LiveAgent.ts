import { observable, action } from 'mobx';

declare global {
  interface Window {
    LIVE_AGENT_AVAILABLE?: boolean;
    _laq?: Array<Function>;

    liveagent?: {
      // Must be called before window.load event
      init: (chatUrl: string, orgId: string, deploymentId: string) => void;

      // Only works on buttonId that has also been registered with showWhenOnline/showWhenOffline
      addButtonEventHandler: (
        buttonId: string,
        callback: (eventType: string) => void
      ) => void;

      // These two must be called before init
      showWhenOnline: (buttonId: string, el: HTMLElement) => void;
      showWhenOffline: (buttonId: string, el: HTMLElement) => void;

      startChat: (buttonId: string) => void;
      addCustomDetail: (
        name: string,
        value: string,
        showToAgent?: boolean
      ) => void;
    };
  }
}

/**
 * Service to manage Salesforceâ€™s LiveAgent external JavaScript.
 */
export default class LiveAgent {
  @observable
  liveAgentAvailable: boolean = !!(
    typeof window !== 'undefined' && window.LIVE_AGENT_AVAILABLE
  );

  _liveAgentButtonId: string = '';

  get liveAgentButtonId(): string {
    return this._liveAgentButtonId;
  }

  set liveAgentButtonId(liveAgentButtonId: string) {
    this._liveAgentButtonId = liveAgentButtonId;

    if (typeof window !== 'undefined') {
      if (window.liveagent) {
        this.listenForLiveAgentEvents(liveAgentButtonId);
      } else {
        if (!window._laq) {
          window._laq = [];
        }

        window._laq.push(() =>
          this.listenForLiveAgentEvents(liveAgentButtonId)
        );
      }
    }
  }

  listenForLiveAgentEvents(liveAgentButtonId: string) {
    window.liveagent!.addButtonEventHandler(
      liveAgentButtonId,
      this.liveAgentEventHandler
    );
  }

  @action.bound
  liveAgentEventHandler(event: string) {
    switch (event) {
      case 'BUTTON_AVAILABLE':
        this.liveAgentAvailable = true;
        break;
      case 'BUTTON_UNAVAILABLE':
        this.liveAgentAvailable = false;
        break;
      default:
        break;
    }
  }
}
