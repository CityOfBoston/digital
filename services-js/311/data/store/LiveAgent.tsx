import { observable, action } from 'mobx';

declare global {
  interface Window {
    /** Set by the setup script. */
    LIVE_AGENT_AVAILABLE?: boolean;

    /**
     * The LiveAgent code automatically executes every function in this array
     * when it loads.
     */
    _laq?: Array<Function>;

    liveagent?: {
      // Must be called before window.load event
      init: (chatUrl: string, orgId: string, deploymentId: string) => void;

      // Only works on buttonId that has also been registered with showWhenOnline/showWhenOffline
      addButtonEventHandler: (
        buttonId: string,
        callback: (eventType: string) => void
      ) => void;

      // These two must be called before init
      showWhenOnline: (buttonId: string, el: HTMLElement) => void;
      showWhenOffline: (buttonId: string, el: HTMLElement) => void;

      startChat: (buttonId: string) => void;
      addCustomDetail: (
        name: string,
        value: string,
        showToAgent?: boolean
      ) => void;
    };
  }
}

/**
 * Service to manage Salesforceâ€™s LiveAgent external JavaScript.
 */
export default class LiveAgent {
  private readonly liveAgentButtonId: string = '';

  @observable
  public liveAgentAvailable: boolean = !!(
    typeof window !== 'undefined' && window.LIVE_AGENT_AVAILABLE
  );

  /**
   * Include this in _document.tsx to put the LiveAgent loading script in place
   * and to add the configuration we need to interact with it.
   */
  public static SetupScript(props: {
    scriptUrl: string;
    buttonId: string;
    chatUrl: string;
    orgId: string;
    deploymentId: string;
  }): JSX.Element {
    return (
      <>
        <script src={process.env.LIVE_AGENT_SCRIPT_SRC} />
        <script
          type="text/javascript"
          dangerouslySetInnerHTML={{
            __html: `
              (function() {
                if (typeof liveagent === 'undefined') {
                  return;
                }

                var buttonId = "${props.buttonId}";
                // We need to make a fake "showWhenOnline" in order to get
                // button events from the server. All of the app showing/hiding
                // is done though addButtonEventHandler because Live Agent does
                // not support adding new buttons after init is called, which
                // runs against what we need in a single-page app.
                liveagent.showWhenOnline(buttonId, document.createElement('DIV'));
                liveagent.addButtonEventHandler(buttonId, function(event) {
                  window.LIVE_AGENT_AVAILABLE = event === 'BUTTON_AVAILABLE';
                });
                liveagent.init(
                  "${props.chatUrl}",
                  "${props.orgId}",
                  "${props.deploymentId}");
              })();
            `,
          }}
        />
      </>
    );
  }

  constructor(liveAgentButtonId: string) {
    this.liveAgentButtonId = liveAgentButtonId;
  }

  public attach() {
    if (!window._laq) {
      window._laq = [];
    }

    if (window.liveagent) {
      this.listenForLiveAgentEvents();
    } else {
      window._laq.push(() => this.listenForLiveAgentEvents());
    }
  }

  private listenForLiveAgentEvents() {
    window.liveagent!.addButtonEventHandler(
      this.liveAgentButtonId,
      this.liveAgentEventHandler
    );
  }

  @action.bound
  private liveAgentEventHandler(event: string) {
    switch (event) {
      case 'BUTTON_AVAILABLE':
        this.liveAgentAvailable = true;
        break;
      case 'BUTTON_UNAVAILABLE':
        this.liveAgentAvailable = false;
        break;
      default:
        break;
    }
  }

  public startChat() {
    window.liveagent!.startChat(this.liveAgentButtonId);
  }
}
