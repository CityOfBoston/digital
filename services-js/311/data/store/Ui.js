// @flow

import { observable, action, computed, reaction } from 'mobx';
import throttle from 'lodash/throttle';

import { HEADER_HEIGHT } from '../../components/style-constants';

// Store to track window and scroll information.
export default class Ui {
  // This tracks the current window height, minus the fixed header height
  @observable visibleHeight: number = Number.MAX_SAFE_INTEGER;
  @observable visibleWidth: ?number;

  @observable scrollY: number = 0;
  @observable debouncedScrollY: number = 0;

  @observable visible: boolean = true;
  @observable reduceMotion: boolean = false;

  debouncedScrollDisposer: ?Function;

  @action
  attach() {
    this.debouncedScrollDisposer = reaction(
      () => this.scrollY,
      (scrollY) => { this.debouncedScrollY = scrollY; },
      {
        name: 'debounced scroll',
        delay: 250,
      },
    );

    if (typeof window === 'undefined') {
      return;
    }

    if (window.matchMedia) {
      this.reduceMotion = window.matchMedia('(prefers-reduced-motion)').matches;
    }

    window.addEventListener('scroll', this.handleScroll);
    window.addEventListener('resize', this.handleResize);

    this.visible = !document.hidden;
    document.addEventListener('visibilitychange', this.handleVisibility);

    this.handleScroll();
    this.handleResize();
  }

  detach() {
    if (this.debouncedScrollDisposer) {
      this.debouncedScrollDisposer();
      this.debouncedScrollDisposer = null;
    }

    if (typeof window === 'undefined') {
      return;
    }

    window.removeEventListener('scroll', this.handleScroll);
    window.removeEventListener('resize', this.handleResize);

    document.removeEventListener('visibilitychange', this.handleVisibility);
  }

  handleScroll = throttle(action('scroll handler', () => {
    this.scrollY = window.scrollY || (document.documentElement && document.documentElement.scrollTop) || 0;
  }), 100);

  handleResize = throttle(action('resize handler', () => {
    this.visibleHeight = window.innerHeight - HEADER_HEIGHT;
    this.visibleWidth = window.innerWidth;
  }), 100);

  @action.bound
  handleVisibility() {
    this.visible = !document.hidden;
  }

  @computed get mediaSmall(): boolean { return !!this.visibleWidth && this.visibleWidth > 480; }
  @computed get mediaMedium(): boolean { return !!this.visibleWidth && this.visibleWidth > 768; }
  @computed get mediaLarge(): boolean { return !!this.visibleWidth && this.visibleWidth > 840; }
  @computed get mediaXLarge(): boolean { return !!this.visibleWidth && this.visibleWidth > 980; }
  @computed get mediaXXLarge(): boolean { return !!this.visibleWidth && this.visibleWidth > 1280; }

  // true if less that mediaLarge, but false if we haven't determined the screen size yet
  @computed get belowMediaLarge(): boolean { return !!this.visibleWidth && !this.mediaLarge; }
}
