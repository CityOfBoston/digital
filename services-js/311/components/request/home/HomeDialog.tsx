import React, { FormEvent } from 'react';
import { action, observable, reaction, runInAction } from 'mobx';
import { observer } from 'mobx-react';
import debounce from 'lodash/debounce';
import Router from 'next/router';

import {
  FetchGraphql,
  GaSiteAnalytics,
  getParam,
} from '@cityofboston/next-client-common';

import { ServiceSummary } from '../../../data/types';
import loadServiceSuggestions from '../../../data/queries/load-service-suggestions';
import loadTopServiceSummaries from '../../../data/queries/load-top-service-summaries';

import FormDialog from '../../common/FormDialog';
import LoadingIcons from '../../common/LoadingIcons';
import { CENTERED_DIALOG_STYLE } from '../../style-constants';
import TranslateDialog from '../translate/TranslateDialog';

import HomePane from './HomePane';
import ChooseServicePane from './ChooseServicePane';
import { GetInitialProps } from '../../../pages/_app';
import { LanguagePreference } from '../../../data/store/BrowserLanguage';
import Ui from '../../../data/store/Ui';
import LiveAgent from '../../../data/store/LiveAgent';

type Stage = 'home' | 'choose';

export type InitialProps = {
  topServiceSummaries: ServiceSummary[];
  description: string;
  stage: Stage;
  bypassTranslateDialog: boolean;
};

export type Props = InitialProps & {
  fetchGraphql: FetchGraphql;
  siteAnalytics: GaSiteAnalytics;
  ui: Ui;
  liveAgent: LiveAgent;
  languages: LanguagePreference[];
};

@observer
export default class HomeDialog extends React.Component<Props> {
  @observable private description: string = '';

  @observable.shallow
  private suggestedServiceSummaries: (ServiceSummary[]) | null = null;
  private lastSuggestedServiceSummariesPromise: Promise<
    ServiceSummary[]
  > | null = null;
  private serviceSuggestionsDisposer: Function | null = null;

  // Called by RequestLayout
  static getInitialProps: GetInitialProps<
    InitialProps,
    'query',
    'fetchGraphql'
  > = async ({ query }, { fetchGraphql }): Promise<InitialProps> => {
    const { stage, description, translate } = query;

    return {
      topServiceSummaries: await loadTopServiceSummaries(fetchGraphql, 5),
      stage: stage === 'choose' ? stage : 'home',
      description: getParam(description, ''),
      bypassTranslateDialog: translate === '0',
    };
  };

  @action
  componentWillMount() {
    this.description = this.props.description;
  }

  componentDidMount() {
    this.serviceSuggestionsDisposer = reaction(
      () => this.description,
      (description: string) => {
        if (description === '') {
          this.suggestedServiceSummaries = [];
        } else {
          this.suggestedServiceSummaries = null;
          this.suggestServices(description);
        }
      },
      {
        fireImmediately: true,
        name: 'update service suggestions from description',
      }
    );

    LoadingIcons.preload();
  }

  componentWillUnmount() {
    if (this.serviceSuggestionsDisposer) {
      this.serviceSuggestionsDisposer();
    }
  }

  suggestServices = debounce(async (description: string) => {
    try {
      this.lastSuggestedServiceSummariesPromise = loadServiceSuggestions(
        this.props.fetchGraphql,
        description
      );

      const suggestedServiceSummaries = await this
        .lastSuggestedServiceSummariesPromise;

      runInAction('suggestServices result', () => {
        this.suggestedServiceSummaries = suggestedServiceSummaries;
      });
    } catch (e) {
      runInAction('suggestServices error', () => {
        this.suggestedServiceSummaries = [];
      });
      throw e;
    }
  }, 500);

  @action.bound
  handleDescriptionChanged(ev: FormEvent<HTMLTextAreaElement>) {
    this.description = ev.currentTarget.value;
  }

  @action.bound
  routeToChoose() {
    const { lastSuggestedServiceSummariesPromise, description } = this;
    const { siteAnalytics } = this.props;

    // We send site analytics only when routing to the choose dialog so that we
    // don't send events for the autocomplete preloading.
    if (lastSuggestedServiceSummariesPromise) {
      lastSuggestedServiceSummariesPromise.then(arr => {
        if (arr.length) {
          siteAnalytics.sendEvent('results', {
            category: 'Prediction',
            label: description,
            value: arr.length,
          });
        } else {
          siteAnalytics.sendEvent('no-results', {
            category: 'Prediction',
            label: description,
            value: 0,
          });
        }
      });
    }

    Router.push(
      `/request?stage=choose&description=${encodeURIComponent(
        this.description.trim()
      )}`,
      `/request?description=${encodeURIComponent(this.description.trim())}`
    ).then(() => window.scrollTo(0, 0));
  }

  render() {
    const { stage, bypassTranslateDialog, languages } = this.props;

    const translateLanguage = TranslateDialog.findLanguage(languages);
    const showTranslate =
      translateLanguage &&
      translateLanguage !== 'en' &&
      !bypassTranslateDialog &&
      stage === 'home';

    const narrow = stage === 'choose';

    if (showTranslate) {
      return <TranslateDialog languages={languages} showContinueInEnglish />;
    }

    return (
      <div className={CENTERED_DIALOG_STYLE}>
        <FormDialog narrow={narrow} noPadding>
          {stage === 'home' && this.renderHome()}
          {stage === 'choose' && this.renderServicePicker()}
        </FormDialog>
      </div>
    );
  }

  renderHome() {
    const { topServiceSummaries, liveAgent, ui } = this.props;
    return (
      <HomePane
        liveAgent={liveAgent}
        ui={ui}
        description={this.description}
        handleDescriptionChanged={this.handleDescriptionChanged}
        nextFn={this.routeToChoose}
        topServiceSummaries={topServiceSummaries}
      />
    );
  }

  renderServicePicker() {
    const { ui, siteAnalytics } = this.props;
    return (
      <ChooseServicePane
        description={this.description}
        suggestedServiceSummaries={this.suggestedServiceSummaries}
        ui={ui}
        siteAnalytics={siteAnalytics}
      />
    );
  }
}
