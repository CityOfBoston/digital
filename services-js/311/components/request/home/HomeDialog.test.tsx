import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { makeServerContext } from '../../../lib/test/make-context';
import HomeDialog, { InitialProps, Props } from './HomeDialog';
import {
  FetchGraphql,
  GaSiteAnalytics,
} from '@cityofboston/next-client-common';
import Ui from '../../../data/store/Ui';
import LiveAgent from '../../../data/store/LiveAgent';

jest.mock('../../../data/queries/load-top-service-summaries');
jest.mock('../../../data/queries/load-service-suggestions');

const loadTopServiceSummaries: jest.Mock = (require('../../../data/queries/load-top-service-summaries') as any)
  .default;
const loadServiceSuggestions: jest.Mock = (require('../../../data/queries/load-service-suggestions') as any)
  .default;

jest.mock('lodash/debounce');
const debounce: jest.Mock = require('lodash/debounce') as any;

const SERVICE_SUMMARIES = [
  {
    name: 'Cosmic Incursion',
    code: 'CSMCINC',
    description: 'Something is threatening the fabric of the universe',
    hasMetadata: true,
    locationRequired: true,
  },
];

beforeEach(() => {
  // mock implementation always fires, but after a delay
  debounce.mockImplementation(fn => (...args) =>
    Promise.resolve().then(fn.bind(null, ...args))
  );
});

describe('home page', () => {
  let initialProps: InitialProps;
  let fetchGraphql: FetchGraphql;

  beforeEach(async () => {
    fetchGraphql = jest.fn();

    loadTopServiceSummaries.mockReturnValue(SERVICE_SUMMARIES);

    const ctx = makeServerContext('/request');
    initialProps = await HomeDialog.getInitialProps(ctx, { fetchGraphql });
  });

  test('getInitialProps', () => {
    expect(initialProps.stage).toEqual('home');
    expect(initialProps.topServiceSummaries).toHaveLength(1);
  });

  test('rendering', () => {
    const component = renderer.create(
      React.createElement<Props>(HomeDialog, {
        fetchGraphql,
        siteAnalytics: new GaSiteAnalytics(),
        ui: new Ui(),
        liveAgent: new LiveAgent(''),
        languages: [],
        topServiceSummaries: initialProps.topServiceSummaries,
        description: initialProps.description,
        stage: initialProps.stage,
        bypassTranslateDialog: initialProps.bypassTranslateDialog,
      })
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('choose page', () => {
  let initialProps: InitialProps;
  let fetchGraphql: FetchGraphql;

  beforeEach(async () => {
    fetchGraphql = jest.fn();

    loadTopServiceSummaries.mockReturnValue(SERVICE_SUMMARIES);

    const ctx = makeServerContext('/request', {
      stage: 'choose',
      description: 'Thanos is attacking',
    });
    initialProps = await HomeDialog.getInitialProps(ctx, { fetchGraphql });
  });

  test('getInitialProps', () => {
    expect(initialProps.stage).toEqual('choose');
    expect(initialProps.description).toEqual('Thanos is attacking');
  });

  test('rendering', () => {
    const component = renderer.create(
      React.createElement<Props>(HomeDialog, {
        fetchGraphql,
        siteAnalytics: new GaSiteAnalytics(),
        ui: new Ui(),
        liveAgent: new LiveAgent(''),
        languages: [],
        topServiceSummaries: initialProps.topServiceSummaries,
        description: initialProps.description,
        stage: initialProps.stage,
        bypassTranslateDialog: initialProps.bypassTranslateDialog,
      }),
      { createNodeMock: () => document.createElement('div') }
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});

// disabled as per Reilly 9/23 jm
xdescribe('integration', () => {
  let wrapper;
  let resolveSuggestions;

  beforeEach(() => {
    loadServiceSuggestions.mockReturnValue(
      new Promise(resolve => {
        resolveSuggestions = resolve;
      })
    );

    wrapper = mount(
      <HomeDialog
        liveAgent={new LiveAgent('')}
        siteAnalytics={new GaSiteAnalytics()}
        ui={new Ui()}
        languages={[]}
        description="Thanos is attacking"
        fetchGraphql={jest.fn()}
        stage="choose"
        topServiceSummaries={[]}
        bypassTranslateDialog
      />
    );
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test('description and suggestions', async () => {
    // get on the other side of the debounced lookup triggered by component will mount
    await Promise.resolve();

    expect(loadServiceSuggestions).toHaveBeenCalledWith(
      expect.anything(),
      'Thanos is attacking'
    );

    await resolveSuggestions(SERVICE_SUMMARIES);
    wrapper.update();

    // The service should now appear in the list
    const serviceLink = wrapper
      .find('a')
      .filterWhere(el => el.text().startsWith('Cosmic Incursion'));
    expect(serviceLink.length).toEqual(1);
    // We look to the parent, which is a Next Link component.
    expect(serviceLink.first().props().href).toEqual('/request/CSMCINC');
  });
});
