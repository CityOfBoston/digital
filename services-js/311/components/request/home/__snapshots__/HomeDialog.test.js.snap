// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`choose page rendering 1`] = `
<div
  className={
    Object {
      "data-css-19tfmgz": "",
    }
  }
>
  <div
    className=" br br-t400 br--y css-1aypqv0"
    style={
      Object {
        "maxWidth": 800,
      }
    }
  >
    <div>
      <div
        className="p-a300 p-a800--xl"
        style={
          Object {
            "paddingBottom": ".75rem",
          }
        }
      >
        <div
          className="sh sh--y"
        >
          <h1
            className="sh-title"
          >
            BOS:311 — Choose a Service
          </h1>
        </div>
        <div
          className="m-v500 t--intro"
        >
          “
          Thanos is attacking
          ”
        </div>
      </div>
      <div
        className="b b--g p-a300 p-a800--xl"
      >
        <div>
          <div
            className="t--info"
          >
            Matching your request to BOS:311 services…
          </div>
          <div
            className="p-a300 g css-rdsogp"
          >
            <div
              className="g--4 css-19eoixt"
            >
              <span
                className="css-17mox4x"
              >
                <svg
                  className={
                    Object {
                      "data-css-1acy7qc": "",
                    }
                  }
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "<use xlink:href=\\"/static/img/svg/loading-icons.svg#311_icons_general_lighting_request\\" height=\\"100%\\"></use>",
                    }
                  }
                  role="img"
                />
              </span>
            </div>
            <div
              className="g--4 css-19eoixt"
            >
              <span
                className="css-17mox4x"
              >
                <svg
                  className={
                    Object {
                      "data-css-1acy7qc": "",
                    }
                  }
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "<use xlink:href=\\"/static/img/svg/loading-icons.svg#311_icons_general_lighting_request\\" height=\\"100%\\"></use>",
                    }
                  }
                  role="img"
                />
              </span>
            </div>
            <div
              className="g--4 css-19eoixt"
            >
              <span
                className="css-17mox4x"
              >
                <svg
                  className={
                    Object {
                      "data-css-1acy7qc": "",
                    }
                  }
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "<use xlink:href=\\"/static/img/svg/loading-icons.svg#311_icons_general_lighting_request\\" height=\\"100%\\"></use>",
                    }
                  }
                  role="img"
                />
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`home page rendering 1`] = `
<div
  className={
    Object {
      "data-css-19tfmgz": "",
    }
  }
>
  <div
    className=" br br-t400 br--y css-1aypqv0"
    style={Object {}}
  >
    <div>
      <div
        className="p-a300 p-a800--xl"
      >
        <div
          className="sh sh--y"
        >
          <h1
            className="sh-title"
          >
            BOS:311 — Request City Services
          </h1>
        </div>
        <div
          className="t--info m-v300"
          style={
            Object {
              "fontStyle": "normal",
            }
          }
        >
          Through BOS:311, you can request non-emergency services from the City.
        </div>
        <div
          className="g m-t500"
        >
          <form
            className="g--7 css-1vm2aiv"
            onSubmit={[Function]}
          >
            <div>
              <h2
                className="stp m-v100"
              >
                What can we help with?
              </h2>
              <div
                className="br br-a100 css-1szg2lt"
                style={
                  Object {
                    "maxHeight": 222,
                    "minHeight": 137,
                  }
                }
              >
                <textarea
                  aria-label="Description of the problem"
                  className={
                    Object {
                      "data-css-zkcm5e": "",
                    }
                  }
                  defaultValue=""
                  name="description"
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onInput={[Function]}
                  placeholder=""
                  style={
                    Object {
                      "height": 137,
                    }
                  }
                />
              </div>
            </div>
            <div
              className="m-t500"
              style={
                Object {
                  "textAlign": "right",
                }
              }
            >
              <span
                style={
                  Object {
                    "position": "relative",
                  }
                }
              >
                <button
                  className="btn css-1q8b8td"
                  disabled={true}
                  type="submit"
                >
                  Start a Request
                </button>
                <span
                  onClick={[Function]}
                  role="presentation"
                  style={
                    Object {
                      "bottom": 0,
                      "left": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                    }
                  }
                />
              </span>
            </div>
          </form>
          <div
            className="g--1 m-v300 css-18t5hgw"
          >
            <div
              className="br br-a150 css-s4kbi"
            />
            <div
              className="br br-a150 t--info css-u5arvv"
            >
              or
            </div>
            <div
              className="br br-a150 css-s4kbi"
            />
          </div>
          <div
            className="g--4"
          >
            <h2
              className="a11y--h"
            >
              Popular Services
            </h2>
            <div
              className="t--info"
              style={
                Object {
                  "fontStyle": "normal",
                }
              }
            >
              You can also make a request for one of these popular services:
            </div>
            <ul
              className="ul m-v300"
            >
              <li
                className="t--info"
              >
                <a
                  className="m-v100"
                  href="/request/CSMCINC"
                  onClick={[Function]}
                >
                  Cosmic Incursion
                </a>
              </li>
            </ul>
            <div
              className="t--info"
            >
              <a
                href="/services"
                onClick={[Function]}
              >
                See all services…
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="b b--g p-a300 p-a800--xl css-17cwx2g"
        role="search"
        style={
          Object {
            "paddingBottom": 0,
            "paddingTop": 0,
          }
        }
      >
        <h2
          className="a11y--h"
        >
          Public Cases
        </h2>
        <div
          className="g"
        >
          <form
            acceptCharset="UTF-8"
            action="/lookup"
            className="sf sf--y sf--md g--7 m-v400"
            method="get"
            onSubmit={[Function]}
          >
            <div
              className="sf-i"
            >
              <input
                aria-label="Search field"
                className="sf-i-f"
                name="q"
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Search by case reference # or keywords…"
                type="text"
                value=""
              />
              <button
                className="sf-i-b"
                type="submit"
              >
                Search
              </button>
            </div>
          </form>
          <div
            className="g--1"
          />
          <div
            className="g--4 t--info css-12alag6"
          >
            <a
              href="/search"
              onClick={[Function]}
            >
              Browse public cases
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
